unique violations

arizzitano opened the pull request:
### What
Simplify and deduplicate the list of sitewide violations. To get the list of unique violations, we index them by WCAG violation code + DOM selector. The index page now displays the new counts, which are significantly smaller and more actionable.

Add a new template to list these unique violations by category (errors, warnings, or notices) and display the number of pages each occurs for. These generated pages (`errors.html`, `warnings.html`, and `notices.html`) are now linked from the index page.

### Why
If a violation occurs for the same element on multiple pages (e.g. some chunk of the site header), we don't want to count it more than once. Grouping violations by code + selector allows us to:
- quickly identify and remediate problematic violations
- easily identify erroneous violations and flag them in the ignore rules
- accurately assess overall WCAG2AA conformance level

### Where
Generated pages are [viewable here](https://build.testeng.edx.org/job/edx-platform-accessibility-pr/30516/artifact/edx-platform/reports/pa11ycrawler/html/index.html). **Please note**: JS hosting is disabled on Jenkins so some of the table styling and functionality will not work, but you can get the general idea.

To run locally (and get the JS functionality), you can checkout the branch `arizzitano/pa11ycrawler-unique` and run the following:
```
paver pa11ycrawler --fasttest --skip-clean --fetch-course --with-html
```
then open `edx-platform/reports/pa11ycrawler/html/index.html` in a browser on your host machine.

### Screenshots
Changes to the index page numbers. Note that I removed the "total" count as it's not really an actionable number. Notices aren't necessarily actionable either (they are more like guidelines) but it's valuable to see them all in one place so we can easily pick out the ones that require ignore rules.
![image](https://cloud.githubusercontent.com/assets/491289/22761173/9c921162-ee27-11e6-882a-fe3575026bb0.png)

Generated errors page:
![image](https://cloud.githubusercontent.com/assets/491289/22761460/c75b37ec-ee28-11e6-926e-6059287d9574.png)

Warnings page:
![image](https://cloud.githubusercontent.com/assets/491289/22761472/d72d1dde-ee28-11e6-9f02-5fe3cde0812f.png)

Notices page:
![image](https://cloud.githubusercontent.com/assets/491289/22761484/ee675546-ee28-11e6-98b6-d2e27bdedf1e.png)

Design feedback is welcome!

/cc @jzoldak @benpatterson @cptvitamin (please feel free to add other folks who might be interested as well)

arizzitano committed:
unique violation counts and summary pages for each type (AC-702)

basic unique violations page

add table for unique counts

add titles and a few fields in generated pages

cleanup render_template args a bit

fix filtering

encode key before hashing

try actually encoding it this time

make template look more like detail view

render different violation types into different pages

better violation count indicator

retool counts on index page

link violation pages from index

change page title wording

explicit reverse sort

test that the violation files are created

jzoldak commented on pa11ycrawler/html.py:
```diff
@@ -91,49 +104,66 @@ def render_html(data_dir, output_dir):
+        render_template(env, unique_path, UNIQUE_TEMPLATE, {
+            "grouped_violations": sorted(
+                grouped_violations[violation_type].values(),
+                key=lambda item: -len(item['pages'])
```
nit:
what does the "-" do? sort descending?
Probably worthwhile to add a comment explaining.

arizzitano replied:
yup, this is hacky... will fix.

jzoldak commented on pa11ycrawler/templates/base.html:
```diff
@@ -45,9 +45,10 @@
                     $('.filterby-btn.current').removeClass('current').attr('aria-pressed', 'false');
                     $(this).addClass('current').attr('aria-pressed', 'true');
-                })
+                });
```
oh snaps! good find.

jzoldak reviewed (approved):
The code looks good.
We should be measuring coverage on this repo. Could you please put in a ticket for that if there isn't one already?
I'm thinking about coverage because I was wondering how much of these changes are covered by tests. Unsure. 
Could you take a peek at how difficult it would be to add something to test_html.py? Even if it's not complete coverage, some addition might be worthwhile. Your call.

arizzitano commented:
@jzoldak thanks for the review! I added some test coverage for the html output and fixed that iffy reverse-sort, and since the code looks good to you I'm going to merge.

I agree that the tests for the html stuff are pretty minimal. I think the best solution is more focused unit tests (since checking the validity of the written reports would basically require more scraping). Unit tests are difficult now because `html.py` is essentially a one-off script where all the functions depend on each other. I'd like to spend some time modularizing it and writing some focused tests, but I'd prefer to punt this to a future release since the scope of this one is already pretty big.
